src/js/playlist/playlist.js

```diff
@@ -30,34 +30,38 @@ define([
         return list;
     };
 
-    /** Filters the sources by taking the first playable type and eliminating sources of a different type **/
-    var _filterSources = Playlist.filterSources = function (sources, androidhls) {
-        var selectedType,
-            newSources = [];
+    //  Choose from the sources a type which matches our most preferred provider
+    function _chooseType(sources) {
+        var m = _.map(sources, function(s) {
+            var provider = Providers.choose(s);
+            var priority = Providers.priority(provider);
 
-        if (!sources) {
-            return;
-        }
+            return {
+                priority : priority,
+                type : s.type
+            };
+        });
+
+        var best = _.max(m, _.property('priority'));
+
+        return best.type;
+    }
 
-        _.each(sources, function (originalSource) {
-            originalSource.androidhls =  androidhls;
-            var source = Source(originalSource);
 
-            if (!source) {
+    // A playlist item may have multiple different sources, but we want to stick with one.
+    var _filterSources = Playlist.filterSources = function (sources, androidhls) {
+        sources = _.compact(_.map(sources, function(originalSource) {
+            if (! _.isObject(originalSource)) {
                 return;
             }
 
-            if (chooseProvider(source)) {
-                // We want sources of all the same type since they may be of different quality levels
-                selectedType = selectedType || source.type;
+            originalSource.androidhls =  androidhls;
```

Comment:
I added androidhls here, but this and other params from the top level config that providers need to know about should be handled in playlist normalization.